services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryManager.java

```diff
@@ -0,0 +1,166 @@
+                                  Function<UploadRequest, Upload> uploadFunction) {
+
+        this.transferConfiguration = transferConfiguration;
+        this.uploadFunction = uploadFunction;
+    }
+
+    public UploadDirectory uploadDirectory(UploadDirectoryRequest uploadDirectoryRequest) {
+        CompletableFuture<CompletedUploadDirectory> returnFuture = new CompletableFuture<>();
+        Path directory = uploadDirectoryRequest.sourceDirectory();
+
+        List<CompletedUpload> uploads = new ArrayList<>();
+        List<FailedUpload> failedUploads = new ArrayList<>();
+        List<CompletableFuture<CompletedUpload>> uploadFutures;
+
+        try (Stream<Path> entries = listFiles(directory, uploadDirectoryRequest)) {
+            uploadFutures =
+                entries.map(path -> {
+                    CompletableFuture<CompletedUpload> future =
+                        uploadSingleFile(uploadDirectoryRequest, uploads, failedUploads, path);
+                    // Forward cancellation of the return future to all individual futures.
+                    CompletableFutureUtils.forwardExceptionTo(returnFuture, future);
+                    return future;
+                }).collect(Collectors.toList());
+        }
+
+        CompletableFuture.allOf(uploadFutures.toArray(new CompletableFuture[0]))
+                         .whenComplete((r, t) -> returnFuture.complete(
+                             DefaultCompletedUploadDirectory.builder()
+                                                            .failedUploads(failedUploads)
+                                                            .successfulUploads(uploads)
+                                                            .build()));
+        return new DefaultUploadDirectory(returnFuture);
+    }
+
+    private CompletableFuture<CompletedUpload> uploadSingleFile(UploadDirectoryRequest uploadDirectoryRequest,
+                                                                List<CompletedUpload> uploads,
+                                                                List<FailedUpload> failedUploads,
+                                                                Path path) {
+        int nameCount = uploadDirectoryRequest.sourceDirectory().getNameCount();
+        UploadRequest uploadRequest = constructUploadRequest(uploadDirectoryRequest, nameCount, path);
+        CompletableFuture<CompletedUpload> future = uploadFunction.apply(uploadRequest)
+                                                                  .completionFuture();
+        future.whenComplete((r, t) -> {
+            if (t != null) {
+                failedUploads.add(DefaultFailedUpload.builder()
+                                                     .exception(t)
+                                                     .path(path)
+                                                     .build());
+            } else {
+                uploads.add(r);
```

Comment:
Do we have some serialization mechanism that i'm missing, or is failedUploads.add and uploads.add a concurrency risk?